fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2694,6 +2706,58 @@ public void addRebuildRecordCountsJob(List<CompletableFuture<Void>> work) {
         work.add(future);
     }
 
+    /**
+     * Return a list of boundaries separating the key ranges maintained by each FDB server. This information can be
+     * useful for splitting a large task (e.g., rebuilding an index for a large record store) into smaller tasks (e.g.,
+     * rebuilding the index for records in certain primary key ranges) more evenly so that they can be executed in a
+     * parallel fashion efficiently. The returned boundaries are an estimate from FDB's locality API and may not
+     * represent the exact boundary locations at any database version.
+     * <p>
+     * The boundaries are returned as a list which is sorted and does not contain any duplicates. The first element of
+     * the list is greater than or equal to <code>low</code>, and the last element is less than or equal to
+     * <code>high</code>.
+     * <p>
+     * This implementation may not work when there are too many shard boundaries to complete in a single transaction.
+     *
+     * @param low low endpoint of primary key range (inclusive)
+     * @param high high endpoint of primary key range (exclusive)
+     * @return the list of boundary primary keys
+     */
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    public RecordCursor<Tuple> getPrimaryKeyBoundariesAsync(@Nonnull Tuple low, @Nonnull Tuple high) {
+        final Transaction transaction = ensureContextActive();
+        byte[] rangeStart = recordsSubspace().pack(low);
+        byte[] rangeEnd = recordsSubspace().pack(high);
+        CloseableAsyncIterator<byte[]> cursor = LocalityUtil.getBoundaryKeys(transaction, rangeStart, rangeEnd);
+        final boolean hasSplitRecordSuffix = hasSplitRecordSuffix();
+        DistinctFilterCursorClosure closure = new DistinctFilterCursorClosure();
+        return new IteratorCursor<>(getExecutor(), cursor)
+                .flatMapPipelined(
+                        result -> new IteratorCursor<>(getExecutor(),
+                                transaction.snapshot().getRange(result, rangeEnd, 1).iterator()),
+                        DEFAULT_PIPELINE_SIZE)
+                .map(keyValue -> {
+                    Tuple recordKey = recordsSubspace().unpack(keyValue.getKey());
+                    return hasSplitRecordSuffix ? recordKey.popBack() : recordKey;
+                })
+                // The input stream is expected to be sorted so this filter can work to de-duplicate the data.
+                .filter(closure::pred);
+    }
+
+    private static class DistinctFilterCursorClosure {
+        private Tuple previousKey = null;
+
+        private boolean pred(Tuple key) {
+            if (key == previousKey) {
```

Comment:
This should be `key.equals(previousKey)`, right?

Reply:
Done. Thanks!

Reply:
Uh, or `TupleHelpers.equals`, I suppose.